File: nodecore-cli/src/main/java/nodecore/cli/serialization/TransactionInfoUnion.java
// VeriBlock NodeCore CLI
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.cli.serialization;

import nodecore.api.grpc.VeriBlockMessages;

public class TransactionInfoUnion {
    public TransactionInfoUnion(final VeriBlockMessages.TransactionUnion union) {
        switch (union.getTransactionCase()) {
            case UNSIGNED:
                type = "unsigned";
                unsigned = new TransactionInfo(union.getUnsigned());
                break;
            case SIGNED:
                type = "signed";
                signed = new SignedTransactionInfo(union.getSigned());
                break;
            case TRANSACTION_NOT_SET:
                type = "unknown";
                break;
        }
    }

    public TransactionInfoUnion(final VeriBlockMessages.TransactionInfoUnion union) {
        switch (union.getTransactionCase()) {
            case UNSIGNED:
                type = "unsigned";
                unsigned = new TransactionInfo(union.getUnsigned().getTransaction());
                break;
            case SIGNED:
                type = "signed";
                signed = new SignedTransactionInfo(union.getSigned());
                break;
            case TRANSACTION_NOT_SET:
                type = "unknown";
                break;
        }
    }

    public String type;

    public TransactionInfo unsigned;

    public SignedTransactionInfo signed;
}

File: nodecore-cli/src/main/java/nodecore/cli/serialization/TransactionInfoUnion.kt
// VeriBlock NodeCore CLI
// Copyright 2017-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.cli.serialization

import nodecore.api.grpc.RpcTransactionInfoUnion
import nodecore.api.grpc.RpcTransactionUnion

class TransactionInfoUnion(
    val type: String? = null,
    val unsigned: TransactionInfo? = null,
    val signed: SignedTransactionInfo? = null
)

fun RpcTransactionUnion.toTransactionInfoUnion() = when (transactionCase) {
    RpcTransactionUnion.TransactionCase.UNSIGNED -> TransactionInfoUnion(
        type = "unsigned",
        unsigned = TransactionInfo(unsigned)
    )
    RpcTransactionUnion.TransactionCase.SIGNED -> TransactionInfoUnion(
        type = "signed",
        signed = signed.toSignedTransactionInfo()
    )
    RpcTransactionUnion.TransactionCase.TRANSACTION_NOT_SET -> TransactionInfoUnion(
        type = "unknown"
    )
    else -> TransactionInfoUnion()
}

fun RpcTransactionInfoUnion.toModel() = when (transactionCase) {
    RpcTransactionInfoUnion.TransactionCase.UNSIGNED -> TransactionInfoUnion(
        type = "unsigned",
        unsigned = TransactionInfo(unsigned.transaction)
    )
    RpcTransactionInfoUnion.TransactionCase.SIGNED -> TransactionInfoUnion(
        type = "signed",
        signed = signed.toModel()
    )
    RpcTransactionInfoUnion.TransactionCase.TRANSACTION_NOT_SET -> TransactionInfoUnion(
        type = "unknown"
    )
    else -> TransactionInfoUnion()
}
